### ISSUE TITLE
Add an UUID to Datafields and Datalists, and take UUIDs into account during Basket ex- and import

### ISSUE THREAD

**Issue by markuslf** [Icinga/icingaweb2-module-director#2696]
## Expected Behavior

Would it be possible for you to implement the missing "uuid" column on `DirectorDatafield` and `DirectorDatalist` (it's already built in for all or most other types like IcingaCommand etc.)?

My goal: I want to be able to ex- and import a Basket which contains uuids.

Import behaviour:
* If UUID exists: Update the object
* If UUID is new: Insert the object

A Basket with UUIDs:
```
{
    "Command": {
        "cmd-...": {
            ...,
            "uuid": "a2318fe7-862a-412c-b6d9-92c6b4b9994d"
        }
    },
    "ServiceTemplate": {
        "tpl-...": {
            ...,
            "uuid": "aba0d85b-3925-49c3-96c0-25d450bab345"
        }
    },
    "Datafield": {
        "1788": {
            "varname": "...",
            ...
            "uuid": "47e8b431-eaae-4136-8a70-350ce0ca1804"
        },
    ...
```

We already had that in our fork of the 1.8.1 director, but with text-based "guid" columns. I don't exactly get your new concept with binary UUIDs in PHP, therefore the request. If it's possible for you to implement that, that would make our fork obsolete.

**Comment by Thomas-Gelf** [comment]
Hi Markus,

TL;DR: yes, this will be a thing.

The main reasoning behind binary UUIDs is search and join performance in the database (16 VS 36 Bytes). Director uses mainly UUIDv4, while in some projects we also started using UUIDv5 for specific use-cases. Those UUIDs can easily be converted to and from their textual representation.

Currently we actively prevent our new UUIDs from being written to the baskets, mainly for compatibility reasons. Rough roadmap:

* add UUIDs to non-Icinga objects (mainly ImportSource, SyncRule, Job)
* add UUIDs to fields and vars
* use UUIDs for relations (inheritance, groups, vars, fields)
* IDs will still remain in place for a while, as I'm aware of setups directly fiddling with them (even if not officially supported, I do not want to break what they're doing)
* write UUIDs to exported baskets, support reading them. Rules similar to what you stated above, with some more tweaks (and optionally some flags for conflict resolution)

Not directly related tasks:

* UUIDs in the activity log (easy to implement, main concern: schema migration could be expensive)
* so-called "Change Sets", allowing for more precise control over what a Basket might apply (example: add an import, instead of replacing the whole template - would preserve user-changes). They're currently in the make for another feature, a Sync Target Hook
* more basket extensions, allowing us to implement something that could be labelled "Vendor Packages"

Cheers,
Thomas

NB: As your fork already uses (text-based) UUIDs, I'd strongly recommend tweaking your schema upgrade scripts in a way, that writes your existing "GUID" (it's technically the very same thing) in their binary representation into those new UUID fields. If your supported DB systems do not ship a method for doing so, stripping the dashes and converting hex to bin does the job on every DB. That way you can preserve your existing GUIDs, and drop the text-based column, once the upstream project satisfies all your needs.

**Comment by markuslf** [comment]
Yes, tweaking our fork schema is easy, we already did that in our test environment. What I stumbled upon is how you implemented UUIDs (I don't fully understand all necessary details, I'm more on the Python side).

Instead of waiting for the roadmap things to happen, low hanging fruits would be
* add UUID to datafields (the right way)
* add UUID to datalists (the right way)

Basket Compatibility: Hm, why? IMHO it would be ok if a 1.10+ basket can't be imported into a 1.9- director. In 1.10+, it would be easy to check for an "uuid" attribute:
* Import: If missing, simply assign a new UUID to the object.
* Export: Always write the UUID attribute, you need it anyway in the future.

What do you think?

**Comment by Thomas-Gelf** [comment]
UUIDs: they're standard UUIDs, the library we're using is `ramsey/uuid`, conversion is as easy as `Uuid::fromString($text)->getBytes()` and `Uuid::fromBytes($binary)->toString()`.

I'll not ship a basket with UUID-support for only part of it's supported objects. As soon as I do so, the next tool developer (even in our company) ships new baskets with pre-generated UUIDs and than complains about partial functionality ;-) So having UUIDs at least for non-Icinga object types with basket support is a must for this feature. That's part of what I meant with compatibility, for the rest I agree with your conclusions. Otherwise I could easily have published UUIDs to our baskets with v1.9/v1.10.

In the meantime for your fork you could search for this piece of code in `IcingaObject::toPlainObject()`:

```php
            // Do not ship ids for IcingaObjects:
            if ($k === $this->getUuidColumn()) {
                continue;
            }
```

Instead of skipping the column, keep it and transform the value to it's textual representation. Similar modifications might be required in other classes implementing the `ExportInterface`. But be careful, you'll face problems in other parts of the code, based on this method. Cloning Objects is most probably one area, where this could become tricky.

Cheers,
Thomas
